src/cloud-api-adaptor/pkg/podnetwork/common.go

```diff
@@ -4,42 +4,270 @@
+	}
+
+	logger.Printf("Primary interface: %s, Gateway address: %s", primaryDev, gw.String())
+
+	return primaryDev, gw, nil
+}
+
+func setupExternalNetwork(hostNS netops.Namespace, hostPrimaryInterface string, podNS netops.Namespace) error {
+
+	// Get the secondary interface details
+	secIface, secAddrCIDR, secRoute, err := getSecondaryInterfaceDetails(hostNS, hostPrimaryInterface)
+	if err != nil {
+		return fmt.Errorf("failed to get secondary interface details: %w", err)
+	}
+
+	// Move the secondary interface to the pod network namespace
+	err = moveInterfaceToNamespace(hostNS, podNS, secIface, secAddrCIDR, secRoute)
+	if err != nil {
+		return err
+	}
+
+	return nil
+
+}
+
+func getInterfaceDetails(ns netops.Namespace, iface string) (
+	addrCIDR netip.Prefix, defRoute *netops.Route, err error) {
+
+	link, err := ns.LinkFind(iface)
+	if err != nil {
+		return netip.Prefix{}, nil, fmt.Errorf("failed to find link %q: %w", iface, err)
+	}
+
+	addrs, err := link.GetAddr()
+	if err != nil {
+		return netip.Prefix{}, nil, fmt.Errorf("failed to get addresses for link %q: %w", link.Name(), err)
+	}
+
+	for _, a := range addrs {
+		if a.Addr().Is4() {
+			addrCIDR = a
+			break
+		}
+	}
+
+	routes, err := ns.RouteList(&netops.Route{Destination: netops.DefaultPrefix})
+	if err != nil {
+		return netip.Prefix{}, nil, fmt.Errorf("failed to get routes on namespace %q: %w", ns.Path(), err)
+	}
+
+	for _, r := range routes {
+		if r.Device == iface && r.Destination.Bits() == 0 && r.Priority < math.MaxInt {
+			defRoute = r
+			break
+		}
+	}
+
+	fmt.Printf("Interface %q IPv4 address %q and default route %v\n", iface, addrCIDR, defRoute)
+
+	return addrCIDR, defRoute, nil
+}
+
+func getSecondaryInterfaceDetails(ns netops.Namespace, primaryInterface string) (
+	secIface string, secAddrCIDR netip.Prefix, secRoute *netops.Route, err error) {
+
+	links, err := ns.LinkList()
+	if err != nil {
+		return "", netip.Prefix{}, nil, fmt.Errorf("failed to list links in namespace %q: %w", ns.Path(), err)
+	}
+
+	for _, link := range links {
+		if isInterfaceFilteredOut(link.Name()) || link.Name() == primaryInterface {
+			continue
+		}
+
+		addr, route, err := getInterfaceDetails(ns, link.Name())
+		if err != nil {
+			return "", netip.Prefix{}, nil, err
+		}
+
+		// The first interface other than the primary having an IPv4 address and a default route
+		// or an IPv4 address in the same subnet as the primary interface with no default route
+		// is considered the secondary interface.
+
+		if !addr.IsValid() {
+			fmt.Printf("Skipping interface %q as it has no valid IPv4 address\n", link.Name())
+			continue
+		}
+
+		if route != nil {
+			fmt.Printf("Secondary interface %q found with IPv4 address %q and route %v\n", link.Name(), addr, route)
+			return link.Name(), addr, route, nil
+		}
+
+		fmt.Printf("Route is nil, checking if %q is in the same subnet as %q\n", addr, primaryInterface)
+		priAddrCIDR, defRoute, err := getInterfaceDetails(ns, primaryInterface)
```

Comment:
Unless I'm missing something, looks like this `getInterfaceDetails(ns, primaryInterface)` could be called outside the loop?

Reply:
Yes it can be outside the loop. I kept it inside the loop to call it only if there is a valid secondary interface.

Reply:
Not sure I follow your point here. Since ns and primaryInterface don’t change inside the loop, and the earlier conditions only skip some iterations, getInterfaceDetails(ns, primaryInterface) still ends up being called unnecessarily on each pass. Maybe I’m missing something, but as far as I can tell, there’s no condition that prevents it from being called repeatedly.